Prevent re-entry in EOF command processing

wavexx opened the pull request:
cmdqueue will run commands when EOF is reached without returning to the
main loop, which is already incorrect.

However, since it needs to ensure the queue is empty, an st_synchronize
call can result in a re-entrant call to get_command, which will
reprocess EOF again. Even if we removed st_synchronize, another command
could be picked by an unsuspecting manage_inactivity() somewhere else.

Short-circuit EOF processing by clearing card.sdprinting immediately.
This is safe to do, since it's already done by printingHasFinished()
called a few lines down. We're just doing this earlier.

This should fix #3549

wavexx committed:
Prevent re-entry in EOF command processing

cmdqueue will run commands when EOF is reached without returning to the
main loop, which is already incorrect.

However, since it needs to ensure the queue is empty, an st_synchronize
call can result in a re-entrant call to get_command, which will
reprocess EOF again. Even if we removed st_synchronize, another command
could be picked by an unsuspecting manage_inactivity() somewhere else.

Short-circuit EOF processing by closing the file early and checking for
the file state early in get_command.

This should fix #3549

leptun commented:
Won’t this cause the lcd to switch to idle mode too quickly? Also, wouldn’t this cause a pause event after eof is reached to fail?

wavexx commented:
The LCD doesn't seem to be an issue, but the timers are checking IS_SD_PRINTING so it is a problem for that :(

I pushed another alternative approach on the same idea. We close the file early, and check for the file status (which is not a bad thing to do here anyway).

The file is correctly reopened when chaining, and I don't see relevant situation where the file state is checked besides in layer1cal

gudnimg commented:
Tested this quickly on my printer. No crash and the file is only closed once 👍

wavexx commented:
Rebased on the MK3 branch

wavexx committed:
Reset sdpos_atomic when starting a new SD print

Fix an incorrect SD offset on the first G-Code command when the second
SD print is started.

wavexx committed:
Switch a few pointers that don't manipute strings to const

wavexx committed:
Do *not* shorten the current command in printer_smodel_check

printer_smodel_check was incorrectly substituting the final " with a
null in the command to simplify the model string comparison, but in
doing so was also corrupting the next pop from the cmdqueue.

We can modify the current strchr_pointer as long as we *don't* change
it's length. This can cause an incorrect extra read from the queue,
resulting in the last command to be completely ignored.

wavexx committed:
Simplify printingHasFinished

Unswitch the call to file.close().

Do not call quickStop(): motion has already completed due to
st_synchronize.
